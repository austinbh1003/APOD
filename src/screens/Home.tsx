import React, { Component } from 'react';
import styled from 'styled-components';
import Footer from '../components/Footer';
import Loader from '../components/Loader';
import Picture from '../components/Picture';
import FullScreenView from '../components/FullScreenView';
import { PictureData } from '../interfaces/state';
import Header from '../components/Header';

const PageWrapper = styled.div`
  position: relative;
`;

const Wrapper = styled.div`
  display: flex;
  flex-direction: column;
  margin: 0 10%;
  justify-content: space-between;
  min-height: 100vh;
`;

const PictureTitle = styled.h3`
  margin: 20px 0;
  font-weight: 700;
  font-size: 30px;
`;

const ContentWrapper = styled.div`
  display: flex;
  flex-direction: row-reverse;
  flex-wrap: wrap;
`;

const Description = styled.p`
  text-align: justify;
  line-height: 20px;
`;

const TextWrapper = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: center;
  margin: 0 10px 20px 10px;
  flex: 6;
  min-width: 50%;
`;

const BackgroundWrapper = styled.div`
  position: absolute;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  overflow: hidden;
  z-index: -1;
`;

const BackgroundCircle = styled.div`
  height: 100vw;
  width: 100vw;
  border-radius: 50vw;
  background: rgba(255, 255, 255, 0.03);
  position: absolute;
  left: -20vw;
  bottom: -40vw;
`;

interface Props {
  loadApod: () => void;
  loading: boolean;
  pictureData: null | PictureData;
  fullScreenViewVisible: boolean;
  closeFullScreenView: () => void;
  openFullScreenView: () => void;
}

class Home extends Component<Props> {

  componentDidMount() {
    this.props.loadApod();
  }

  render() {
    const { loading, pictureData, fullScreenViewVisible, openFullScreenView, closeFullScreenView } = this.props;
    const imgUrl = pictureData ? pictureData.url : "";
    const title = pictureData ? pictureData.title : "";
    const description = pictureData ? pictureData.explanation : "";
    return (
      <PageWrapper>
        <BackgroundWrapper>
          <BackgroundCircle />
        </BackgroundWrapper>
        {fullScreenViewVisible && <FullScreenView imgSrc={imgUrl} closeFullScreenView={closeFullScreenView} />}
        <Wrapper>
          <Header />
          {loading ? <Loader /> : (
            <ContentWrapper>
              <Picture src={imgUrl} onClick={() => openFullScreenView()} />
              <TextWrapper>
                <PictureTitle>{title}</PictureTitle>
                <Description>{description}</Description>
              </TextWrapper>
            </ContentWrapper>
          )}
          <Footer />
        </Wrapper>
      </PageWrapper>
    )
  }
}

export default Home;