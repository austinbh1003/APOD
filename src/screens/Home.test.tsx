import React from 'react';
import Home from './Home';
import renderer from 'react-test-renderer';

jest.mock('../components/Header', () => 'Header');
jest.mock('../components/Footer', () => 'Footer');
jest.mock('../components/Loader', () => 'Loader');
jest.mock('../components/Picture', () => 'Picture');
jest.mock('../components/Video', () => 'Video');
jest.mock('../components/FullScreenView', () => 'FullScreenView');

describe('Home screen', () => {

  const defaultProps = {
    loadApod: jest.fn(),
    loading: true,
    pictureData: null,
    fullScreenViewVisible: false,
    closeFullScreenView: jest.fn(),
    openFullScreenView: jest.fn()
  }

  it('renders correctly while loading', () => {
    const tree = renderer
      .create(<Home {...defaultProps} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders correctly with a video loaded', () => {
    const props = {
      ...defaultProps,
      loading: false,
      pictureData: {
        date: "2019-05-20",
        explanation: "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.",
        media_type: "video",
        service_version: "v1",
        title: "Planets of the Solar System: Tilts and Spins",
        url: "https://www.youtube.com/embed/my1euFQHH-o?rel=0"
      }
    }
    const tree = renderer
      .create(<Home {...props } />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("renders correctly with an image loaded", () => {
    const props = {
      ...defaultProps,
      loading: false,
      pictureData: {
        copyright: "Po Tato",
        date: "2019-05-21",
        explanation: "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.",
        hdurl: "https://example.com/hdImage.jpg",
        media_type: "image",
        service_version: "v1",
        title: "Sample title",
        url: "https://example.com/image.jpg"
      }
    };
    const tree = renderer
      .create(<Home {...props } />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
