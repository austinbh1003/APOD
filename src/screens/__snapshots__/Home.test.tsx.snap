// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home screen renders correctly while loading 1`] = `
<div
  className="sc-bdVaJa EXKLM"
>
  <div
    className="sc-bZQynM cnTDcC"
  >
    <div
      className="sc-gzVnrw gQCyHa"
    />
  </div>
  <div
    className="sc-bwzfXH hYfLmM"
  >
    <Header />
    <Loader />
    <Footer />
  </div>
</div>
`;

exports[`Home screen renders correctly with a video loaded 1`] = `
<div
  className="sc-bdVaJa EXKLM"
>
  <div
    className="sc-bZQynM cnTDcC"
  >
    <div
      className="sc-gzVnrw gQCyHa"
    />
  </div>
  <div
    className="sc-bwzfXH hYfLmM"
  >
    <Header />
    <div
      className="sc-bxivhb kJYdCf"
    >
      <Video
        src="https://www.youtube.com/embed/my1euFQHH-o?rel=0"
      />
      <div
        className="sc-EHOje iBowpJ"
      >
        <h3
          className="sc-htpNat ZIhSq"
        >
          Planets of the Solar System: Tilts and Spins
        </h3>
        <p
          className="sc-ifAKCX gmAkgN"
        >
          Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.
        </p>
      </div>
    </div>
    <Footer />
  </div>
</div>
`;

exports[`Home screen renders correctly with an image loaded 1`] = `
<div
  className="sc-bdVaJa EXKLM"
>
  <div
    className="sc-bZQynM cnTDcC"
  >
    <div
      className="sc-gzVnrw gQCyHa"
    />
  </div>
  <div
    className="sc-bwzfXH hYfLmM"
  >
    <Header />
    <div
      className="sc-bxivhb kJYdCf"
    >
      <Picture
        onClick={[Function]}
        src="https://example.com/image.jpg"
      />
      <div
        className="sc-EHOje iBowpJ"
      >
        <h3
          className="sc-htpNat ZIhSq"
        >
          Sample title
        </h3>
        <p
          className="sc-ifAKCX gmAkgN"
        >
          Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.
        </p>
      </div>
    </div>
    <Footer />
  </div>
</div>
`;
